src/lib/snarky_pairing/miller_loop.ml

```diff
@@ -0,0 +1,140 @@
+module type Inputs_intf = sig
+  module Impl : Snarky.Snark_intf.S
+
+  module Fq :
+    Snarky_field_extensions.Intf.S
+    with module Impl = Impl
+     and type 'a A.t = 'a
+     and type 'a Base.t_ = 'a
+
+  module Fqe :
+    Snarky_field_extensions.Intf.S_with_primitive_element
+    with module Impl = Impl
+     and module Base = Fq
+
+  module Fqk : sig
+    include
+      Snarky_field_extensions.Intf.S
+      with module Impl = Impl
+       and type 'a Base.t_ = 'a Fqe.t_
+       and type 'a A.t = 'a * 'a
+
+    val special_mul : t -> t -> (t, _) Impl.Checked.t
+
+    val special_div : t -> t -> (t, _) Impl.Checked.t
+  end
+
+  module G1_precomputation :
+    G1_precomputation.S
+    with module Impl = Impl
+     and type 'a Fqe.Base.t_ = 'a Fqe.Base.t_
+     and type 'a Fqe.A.t = 'a Fqe.A.t
+
+  module G2_precomputation :
+    G2_precomputation.S
+    with module Impl = Impl
+     and type 'a Fqe.Base.t_ = 'a Fqe.Base.t_
+     and type 'a Fqe.A.t = 'a Fqe.A.t
+
+  module N : Snarkette.Nat_intf.S
+
+  module Params : sig
+    val loop_count : N.t
+
+    val loop_count_is_neg : bool
+  end
+end
+
+module Make (Inputs : Inputs_intf) = struct
+  open Inputs
+  open Impl
+  open Let_syntax
+
+  let double_line_eval (p : G1_precomputation.t)
+      (c : G2_precomputation.Coeff.t) : (Fqk.t, _) Checked.t =
+    with_label __LOC__
+      (let px, _ = p.p in
+       let%map c1 =
+         let gamma_twist = Fqe.mul_by_primitive_element c.gamma in
+         let%map t = Fqe.mul_field gamma_twist px in
+         Fqe.(c.gamma_x - c.ry - t)
```

Comment:
the unchecked impl in libsnark does the equivalent of `Fqe.(c.gamma_x - t - c.ry)`, seems like those should be equivalent, but just wanted to point it out